nano/node/bootstrap/upward/bootstrapper.cpp

```diff
@@ -0,0 +1,156 @@
+                                                     nano::logger_mt & logger_a,
+                                                     nano::thread_pool & thread_pool_a) :
+store{ store_a },
+peer_manager { peer_manager_a },
+network_constants{ network_constants_a },
+logger{ logger_a },
+thread_pool { thread_pool_a },
+mutex{ }
+{
+	debug_assert (!store.init_error ());
+}
+
+void nano::bootstrap::upward::bootstrapper::start ()
+{
+    // if already started, issue an error
+
+    boot (thread_pool);
+}
+
+void nano::bootstrap::upward::bootstrapper::pause ()
+{
+    // if already paused or stopped, issue an error
+}
+
+void nano::bootstrap::upward::bootstrapper::resume ()
+{
+    // if stopped or not paused, issue an error
+}
+
+void nano::bootstrap::upward::bootstrapper::stop ()
+{
+    // if already stopped, issue an error
+}
+
+bool nano::bootstrap::upward::bootstrapper::is_running () const
+{
+    // return true if both started and not paused
+
+    return true;
+}
+
+void nano::bootstrap::upward::bootstrapper::follow_account (nano::account const & account_a)
+{
+    nano::lock_guard<nano::mutex> lock{ mutex };
+
+	if (!recently_followed_accounts.count (account_a))
+	{
+		accounts_to_follow.push_front (account_a);
+	}
+}
+
+void nano::bootstrap::upward::bootstrapper::boot (nano::thread_pool & thread_pool_a)
+{
+    // the "10min" value should be a combination of config value (default),
+    // but also dynamic, depending on the size of accounts_to_follow, etc
+
+    using namespace std::chrono_literals;
+    thread_pool_a.add_timed_task (std::chrono::steady_clock::now () + 10min,
+                                  [&thread_pool_a, this] ()
+                                  {
+                                      boot_impl ();
+                                      boot (thread_pool_a);
+                                  });
+}
+
+void nano::bootstrap::upward::bootstrapper::boot_impl ()
+{
+    // this method is the core of the bootstrapper; it iterates the accounts_to_follow container and
+    // tries to download more blocks off the network for each of those accounts.
+    //
+    // accounts should make their way into the container by means of bootstrapper::follow_account.
+    //
+    // presumably, the (sole?) caller of follow_account should be an observer that is subscribed to the
+    // block_has_been_confirmed notification (confirmation_height_processor?).
+    //
+    // the observer should check if the confirmed block was a SEND one,
+    // and if it was then ask the bootstrapper to follow_account(confirmed_block->destination()).
+
+    const auto transaction = store.tx_begin_read ();
+
+    nano::lock_guard<nano::mutex> lock{ mutex };
+    while (!accounts_to_follow.empty ())
+    {
+        const auto & account = accounts_to_follow.back ();
+        recently_followed_accounts.emplace (account);
+
+        nano::account_info account_info{};
+        if (!store.account.get (transaction, account, account_info))
```

Comment:
The account being followed won't always be findable, e.g. following the genesis account to accounts that are destinations from genesis transactions.

Reply:
Makes sense, I updated the code so that `account_info`s are now `std::optional` and assume that if we can't find the account in the storage then we'll be expecting an open block as the first block received for that account.